samples/js/photos/src/photo.js

```diff
@@ -0,0 +1,94 @@
+// @flow
+
+// Copyright 2016 Attic Labs, Inc. All rights reserved.
+// Licensed under the Apache License, version 2.0:
+// http://www.apache.org/licenses/LICENSE-2.0
+
+import type {NomsPhoto, PhotoSize} from './types.js';
+
+import type {MapEntry} from '@attic/noms';
+import {
+  StructMirror,
+  equals,
+  invariant,
+  notNull,
+} from '@attic/noms';
+export type SizeEntry = MapEntry<PhotoSize, string>;
+
+/**
+ * createPhoto asynchronously derives a Photo instance from a NomsPhoto.
+ */
+export async function createPhoto(path: string, p: NomsPhoto): Promise<Photo> {
+  const sizeEntries = [];
+  return p.sizes.forEach((url, size) => {
+    sizeEntries.push([size, url]);
+  }).then(() => new Photo(path, p, sizeEntries));
+}
+
+/**
+ * Photo is a wrapper around the NomsPhoto struct.
+ */
+export default class Photo {
+  path: string;
+  nomsPhoto: NomsPhoto;
+
+  _sizeEntries: SizeEntry[];
+
+  constructor(path: string, nomsPhoto: NomsPhoto, sizeEntries: SizeEntry[]) {
+    this.path = path;
+    this.nomsPhoto = nomsPhoto;
+    this._sizeEntries = sizeEntries;
+  }
+
+  equals(p: Photo): boolean {
+    return equals(this.nomsPhoto, p.nomsPhoto);
+  }
+
+  /**
+   * getBestSize gets the best photo that fits in dimensions (width, height),
+   * where best is the smallest photo which is larger than the dimensions, or
+   * failing that the widest - and one presumes therefore the largest.
+   */
+  getBestSize(width: number, height: number): SizeEntry {
+    const bestWidth = this._smallestImageAtLeast('width', width);
```

Comment:
Ok, spent some time on this and it's not as easy as just picking the best size. We need to update all the rendering code because it assumes that the sizes it gets from getBestSize match the real sizes in the photos - which it doesn't any more.

I think we should just accept sideways photos for now, and fix them some other way, whether that is adding code everywhere to deal with sideways photos, or something else.